[RNMobile] Search Block: Accessibility

AmandaRiu opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

### Gutenberg Mobile PR
https://github.com/wordpress-mobile/gutenberg-mobile/pull/3210

**Note**: ~[The PR to add styling to the Search Block](https://github.com/WordPress/gutenberg/pull/30052) must be merged before this PR can be taken out of draft.~


## Description
<!-- Please describe what you have changed or added -->
This PR is one in a series of PRs to port the search block to mobile. This PR implements accessibility support for VoiceOver and TalkBack including:
- Allow screenreader access to Search Block child components only when the block is actively selected. 
- Wrapped each child component in a `View` and added custom labels to make it clear which part of the search block is actively focused and what it's current value is. 
- Screen reader will skip the Search block button if the block is currently set to use an icon only button.

**Known issue**: When double-tapping to activate the Search block on iOS, the placeholder text input field immediately gains focus. I'm not sure why this is or how to fix it. Ideas welcome 😄 

### Breakdown of the accessibility text for each child component:
**Search Label**
- Label: "Search block label. Current value is `<label>`"
- Hint: "Double tap to edit label text"

**Search Input Field**
- Label: "Search input field. Current placeholder value is `<placeholder>`"
- Hint: "Double tap to edit placeholder text"

**Search Button (text button only)**
- Label: "Search button. Current value is `<buttonText>`"
- Hint: "Double tap to edit button text"

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
Tested on Android Pixel 4 running Android 11 using TalkBack and iPhone 11 running iOS version 13.7 using VoiceOver. Testing should include the following scenarios:
- Block not actively selected. Child components should not gain screenreader focus when swiping through items on the screen. 
- Block actively selected. Swipe through each component and verify accessibility label and hint. 

## Screenshots <!-- if applicable -->

### TalkBack recording

https://user-images.githubusercontent.com/5810477/112408169-b19e2000-8ced-11eb-82ad-14751c931c14.mp4

### VoiceOver recording


https://user-images.githubusercontent.com/5810477/112408185-b95dc480-8ced-11eb-9096-330f7d459d02.mp4



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature currently only available in devOnly mode. No changes were made that would effect the web-version of the search block, nor any other block.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Commit:
Add search block to inserter

Only adds a shell of a block.

Commit:
Add label and icon toggle toolbar buttons

Commit:
Add button position dropdown menu

Commit:
Add temporary alert to make toolbar testing easier

This alert will be removed once styling is added in a later PR.

Commit:
Add initial UI and basic styling

Commit:
Move text button to separate file

This will keep the code cleaner and easier to follow.

Commit:
Add basic button style

Commit:
Add classnames and consolidate button code

Commit:
Remove iconButtonOnly alert message

This toolbar action is now visible in the UI so the temporary alert message
is no longer needed.

Commit:
Add platform-specific styles and basic styles

Commit:
Remove flex settings for label

Was causing issues with layout in iOS. Styling will be addressed fully
 in a different PR.

Commit:
Add button position styles; remove temp alert

Commit:
Clean up styles

Commit:
Merge branch 'rnmobile/3073-search-block-basic-ui' into rnmobile/3073-search-block-button-position

Commit:
Consolidate border styling, fix auto-sizing

Commit:
Fix broken button style

AmandaRiu committed:
Merge branch 'master' into rnmobile/3073-search-block-basic-ui

Commit:
Merge branch 'rnmobile/3073-search-block-basic-ui' into rnmobile/3073-search-block-button-position

Commit:
Move CSS_UNITs to shared utils file

Moved the CSS_UNIT code to a shared Utils file in preparation for using the same code inside the native implementation of the search block.

Commit:
Add basic settings panel with UnitControl

Commit:
Add event stubs and UnitControl

Commit:
Remove unused code and set defaults

Commit:
Finish changeUnit handler and remove debugging

Commit:
Configure defaults, min and max widths

Also moved the default values from the web version to the shared utils file.

AmandaRiu committed:
Merge branch 'trunk' into rnmobile/3073-search-block-button-position

Commit:
Merge trunk into branch

Commit:
Merge branch 'trunk' into rnmobile/3073-search-block-settings

# Conflicts:
#	packages/block-library/src/search/edit.native.js
#	packages/block-library/src/search/rich-text.android.scss
#	packages/block-library/src/search/rich-text.ios.scss
#	packages/block-library/src/search/style.native.scss

Commit:
Merge branch 'rnmobile/3073-search-block-button-position' into rnmobile/3073-search-block-settings

Commit:
Merge branch 'trunk' into rnmobile/3073-search-block-settings

# Conflicts:
#	packages/block-library/src/search/edit.native.js

Commit:
Fix focus issues on Search Block

- Set the onFocus() method to ensure if the placeholder text is clicked on when the block is NOT selected then the block will be selected.
- Set isSelected in RichText elements to false for initial draw to prevent odd behavior on iOS where the Search block will attempt to steal focus from other text blocks.
- Set the placeholder text to focused when the block is first added.

Commit:
Add button settings panel w/ button positions

Commit:
Add label for selected button position option

Commit:
Move search label toggle to settings panel

Commit:
Move button icon option to settings panel

Commit:
Remove old now unused code

Aww…it’s so sad to delete so much work. **sniff**.

Commit:
Cleanup use of useRef

Commit:
Track the state of which text is selected

Track which of the RichText components are currently selected to avoid focus issues on iOS.

Commit:
Only draw toolbar controls if block is selected

Commit:
Set RichText identifiers and tagNames

Commit:
Clear selection state when the block is unselected

Commit:
Merge branch 'rnmobile/search-block-text-focus-bugfixes' into rnmobile/search-block-move-toolbar-options

# Conflicts:
#	packages/block-library/src/search/edit.native.js

Commit:
Remove the `width` block setting

The latest designs removed the `width` setting for the search block for now.

Commit:
Modify setting to hide search label field

This is to match the i2 design changes

Commit:
Merge branch 'master' into rnmobile/3073-search-block-basic-ui

Commit:
Make text lose focus when block loses focus

Fixes an issue where if the label or button text was selected in the search block and then a different block was selected, the cursor would still be active in the search block text field.

Commit:
Add check for textInput reference existance before manipulation

This isn't really a concern for mobile, but on web there is an option for `button-only` so this
change ensures that we don't try to manipulate a reference that would be null.

Commit:
Rename textInput reference for clarity

Commit:
Ensure textInput loses focus when block loses focus

AmandaRiu committed:
Update code comment to remove duplicate wording

Co-authored-by: Paul Von Schrottky <paul.von.schrottky@automattic.com>

Commit:
Merge branch 'rnmobile/search-block-text-focus-bugfixes' into rnmobile/search-block-move-toolbar-options

AmandaRiu committed:
Merge branch 'trunk' into rnmobile/search-block-move-toolbar-options

Commit:
Fix issues with merge from trunk

Commit:
Consolidate RichText styles into main style file

Commit:
Use Icon instead of Button for better control of icon coloring

Commit:
Fix width issues for the search button on Android

Commit:
Add min-height to search input to prevent collapsing when no-button selected

Commit:
Clean up border styles -- initial changes

Commit:
Add placeholder text color

Commit:
Update search block styles

- Added light/dark mode support where needed
- Added cursor styling (seems to only work on ios)
- Tweeked plain text placeholder styling

Commit:
Removed custom cursor style

After speaking with design it was decided for the label and placeholder to just keep it consistent with the rest of the blocks.

Commit:
Simplify styles

Commit:
Fix weird corner pixelation in light mode on ios

Commit:
Fix border radius on text input

Commit:
Standardize colors with color studio

Commit:
Merge branch 'trunk' into rnmobile/search-block-styling

Commit:
Merge branch 'trunk' into rnmobile/search-block-styling

Commit:
Simplify styles

Commit:
Use usePreferredColorSchemeStyle directly instead of HOC

Commit:
Condense and simplify style creation

Commit:
Control when screenreaders can access child components

If the block is not currently selected, then the screenreaders ignore children. Once the block has been activated the user may then move between block components to update individual items.

Commit:
Add accessibility support to the search input field

Commit:
Add accessibility support for button and label

Commit:
Icon only button should not be accessible

dcalhoun commented on packages/base-styles/_colors.native.scss:
```diff
@@ -66,12 +66,15 @@ $button-fallback-bg: #595959;
 $gray-50: #646970;
 $gray-60: #50575e;
 $gray-70: #3c434a;
+$gray-80: #2c3338;
```
I ask largely for my own education, but it seems odd that these variables were not present already. Why is that?

AmandaRiu replied:
I'm not sure. I figure it's because they hadn't been used in code yet.

dcalhoun commented on packages/block-library/src/search/edit.native.js:
```diff
@@ -154,112 +155,165 @@ export default function SearchEdit( {
+				) }
+				accessibilityLabel={ `${ __(
+					'Search input field. Current placeholder value is'
+				) } ${ placeholder }` }
```
I found this label to sound quite odd when no placeholder value is set. What do you think about modifying this to be a bit more descriptive for when the placeholder is empty?

Additionally, the a11y labels and hints alternate between using the term "value" and "text." Do you think it'd be clearer if we used one or the other? This would apply to all the a11y labels in this PR.

```js
const placeholderA11yLabel = placeholder
  ? `Current placeholder text is ${placeholder}.`
  : "No placeholder text set.";
// ...
<View accessibilityLabel={__(`Search input field. ${placeholderA11yLabel}`)} />
```

AmandaRiu replied:
Great feedback, thanks! Implemented in 61e7b0ca3895530cd171b4003b4b5558863a150f

dcalhoun commented on packages/block-library/src/search/edit.native.js:
```diff
@@ -154,112 +155,165 @@ export default function SearchEdit( {
+					'Search input field. Current placeholder value is'
+				) } ${ placeholder }` }
+			>
+				<PlainText
```
>**Known issue**: When double-tapping to activate the Search block on iOS, the placeholder text input field immediately gains focus. I'm not sure why this is or how to fix it. Ideas welcome 😄 

It would appear that the `PlainText` component includes [code to request focus on mount](https://github.com/WordPress/gutenberg/blob/956d55682aa49c715f75cdf580d46e0d1c0ebd60/packages/block-editor/src/components/plain-text/index.native.js#L23-L37). I didn't look into the history as to _why_ this is the case, but depending on why it was originally added, we could possibly remove the focus request or add a prop to have it conditionally run.

Additionally, it looks like the `RichText` component used for Search label has [similar code](https://github.com/WordPress/gutenberg/blob/956d55682aa49c715f75cdf580d46e0d1c0ebd60/packages/rich-text/src/component/index.native.js#L726-L739) that already contains conditional logic and a comment with more historical context.

AmandaRiu replied:
I thought the same thing but if I add that same code it makes no difference. I think just like the other comment below, that iOS recognizes this component as an input field and forces focus on it's own. I even tried deleting the `componentDidMount` code in the `PlainText` block and tested again and it will still steal the focus the moment the block is activated.

dcalhoun commented on packages/block-library/src/search/edit.native.js:
```diff
@@ -154,112 +155,165 @@ export default function SearchEdit( {
+	};
+
+	return (
+		<View
```
I found moving focus to this element on iOS announces "Search block, row 53, button, **actions available**." I'm unsure why "actions available" is announced or where it originates. Do you have any insight into this?

AmandaRiu replied:
I spent quit a bit of time trying to figure this one out and I believe it's due to the use of `<PlainText/>`. The `Code` block using this component as well and it also announces "actions available". The `File` block uses it too, but the `PlainText` component has the `editable={false}` property set so it doesn't announce "actions available". If I do the same thing on my block that announcement also goes away so I'm guessing this is VoiceOver recognizing that a native editable component is a child of the Search block.

dcalhoun commented on packages/block-library/src/search/edit.native.js:
```diff
@@ -154,112 +155,165 @@ export default function SearchEdit( {
+			{ isSelected && controls }
+
+			{ showLabel && (
+				<View
+					accessible={ true }
+					accessibilityRole="none"
+					accessibilityHint={ __( 'Double tap to edit label text' ) }
+					accessibilityLabel={ `${ __(
+						'Search block label. Current value is'
+					) } ${ label }` }
+				>
```
I found on Android that when stepping through elements within a selected Search block, it would step through the wrapping element _and_ the input itself (i.e. 6 stops rather than 3). This did not occur on iOS. Did you experience this?

I'm not sure why difference between platforms would occur, but the multiple steps on Android feels a bit odd. Is the wrapping element necessary? I wonder if we could provide helpful custom labels/hints without the wrapping element?

AmandaRiu replied:
I agree, it's really annoying. Unfortunately customizing the `accessibilityLabel` and `accessibilityHint` only seems to work on container elements like `<View/>`. If I move those options directly to the element for editing, it no longer works.

dcalhoun replied:
@AmandaRiu ah, thank you for providing that context. Relatedly, I found https://github.com/facebook/react-native/issues/26739 which appears to capture the root issue. If it is ever resolved, then we may be able to address the duplicative stops issue by removing the superfluous wrapping elements.

dcalhoun reviewed (commented):
This is looking good! Thanks for investing in a11y features. 💯

Commit:
Update accessibility labels for consistency

- Use ‘text’ instead of ‘value’ for consistency
- Use a different message when no placeholder text is set

AmandaRiu commented:
@dcalhoun Thank you so much for the review! I've addressed all your comments. PR is ready for another review round 👍

Commit:
Merge branch 'trunk' into rnmobile/search-block-accessibility-fixes

dcalhoun reviewed (approved):
Great work!
